### PR TITLE
Add option to set a different set of languages allowed to submit per …

### PR DESCRIPTION
**Pull request by meisterT**
…contest.

If empty, falls back to the global definition.

This is part of #2521.

### CODE REVIEW

File: webapp/migrations/Version20241122144232.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20241122144232 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return '';
```
Comment (nickygerritsen) [inline webapp/migrations/Version20241122144232.php:17]: Missing description

File: webapp/migrations/Version20241122144232.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql('CREATE TABLE contestlanguage (cid INT UNSIGNED NOT NULL COMMENT \'Contest ID\', langid VARCHAR(32) NOT NULL COMMENT \'Language ID (string)\', INDEX IDX_ADCB43234B30D9C4 (cid), INDEX IDX_ADCB43232271845 (langid), PRIMARY KEY(cid, langid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB');
+        $this->addSql('ALTER TABLE contestlanguage ADD CONSTRAINT FK_ADCB43234B30D9C4 FOREIGN KEY (cid) REFERENCES contest (cid) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE contestlanguage ADD CONSTRAINT FK_ADCB43232271845 FOREIGN KEY (langid) REFERENCES language (langid) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE executable DROP zipfile');
```
Comment (nickygerritsen) [inline webapp/migrations/Version20241122144232.php:26]: Unrelated.

File: webapp/templates/jury/contest.html.twig [hunk 3]
```diff
@@ -193,6 +193,23 @@
...
+                    <td>
+                        {% if contest.languages is empty %}
+                            <em>all globally enabled languages:</em>
+                            {% set allowedLanguages = languages %}
+                        {% else %}
+                            {% set allowedLanguages = contest.languages %}
+                        {% endif %}
+                        <ul>
+                            {% for language in allowedLanguages %}
+                                <li>{{ language.name }}</li>
```
Comment (nickygerritsen) [inline webapp/templates/jury/contest.html.twig:207]: Do we want to turn this into links?

**Review by nickygerritsen — APPROVED** [review]
There is also one phpcs annotation.

File: webapp/src/Service/SubmissionService.php [hunk 4]
```diff
@@ -453,9 +453,10 @@ public function submitSolution(
             throw new BadRequestHttpException('Submissions for contest (temporarily) disabled');
         }
 
-        if (!$language->getAllowSubmit()) {
+        $allowedLanguages = $this->dj->getAllowedLanguagesForContest($contest);
+        if (!in_array($language, $allowedLanguages, true)) {
```
Comment (eldering) [inline webapp/src/Service/SubmissionService.php:457]: Nitpick, maybe consider:
```suggestion
        if (!in_array($language, $allowedLanguages, strict: true)) {
```
